File: python-webrtc/python/webrtc/functions/get_user_media.py
#
#  Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
#
#  Use of this source code is governed by a BSD-style license
#  that can be found in the LICENSE.md file in the root of the project.
#

from typing import TYPE_CHECKING

import wrtc

from webrtc import MediaStream

if TYPE_CHECKING:
    import webrtc


# TODO constraints
def get_user_media(constraints=None) -> 'webrtc.MediaStream':
    """Method prompts the user for permission to use a media input which produces a :obj:`webrtc.MediaStream`
    with tracks containing the requested types of media.

    Returns:
        :obj:`webrtc.MediaStream`: A :obj:`webrtc.MediaStream` object representing the media stream.
    """
    return MediaStream._wrap(wrtc.getUserMedia())


#: Alias for :func:`get_user_media`
getUserMedia = get_user_media

File: python-webrtc/cpp/src/functions/get_user_media.cpp
//
// Copyright 2022 Il`ya (Marshal) <https://github.com/MarshalX>. All rights reserved.
//
// Use of this source code is governed by a BSD-style license
// that can be found in the LICENSE.md file in the root of the project.
//

#include "../interfaces/media_stream.h"

namespace python_webrtc {

  static MediaStream *GetUserMedia() {
    auto factory = PeerConnectionFactory::GetOrCreateDefault();
    auto stream = factory->factory()->CreateLocalMediaStream(rtc::CreateRandomUuid());

    // TODO (MarshalX) get from bound MediaStreamConstraints
    // https://github.com/MarshalX/python-webrtc/issues/169
    auto audio = true;
    auto video = false;

    if (audio) {
      cricket::AudioOptions options;
      auto source = factory->factory()->CreateAudioSource(options);
      auto track = factory->factory()->CreateAudioTrack(rtc::CreateRandomUuid(), source);
      stream->AddTrack(track);
    }

    if (video) {
//      TODO (MarshalX) create RTCVideoTrackSource
//      https://github.com/MarshalX/python-webrtc/issues/170
//      auto source = ... RTCVideoTrackSource()
//      auto track = factory->factory()->CreateVideoTrack(rtc::CreateRandomUuid(), source);
//      stream->AddTrack(track);
    }

    return MediaStream::holder()->GetOrCreate(factory, stream);
  }

} // namespace python_webrtc
